cpp/src/arrow/dataset/dataset.h

```diff
@@ -43,7 +43,7 @@ namespace dataset {
 /// Note that Fragments have well defined physical schemas which are reconciled by
 /// the Datasets which contain them; these physical schemas may differ from a parent
 /// Dataset's schema and the physical schemas of sibling Fragments.
-class ARROW_DS_EXPORT Fragment {
+class ARROW_DS_EXPORT Fragment : public std::enable_shared_from_this<Fragment> {
```

Comment:
A further improvement would be to also expose `std::string Fragment::name()` or similar for logging purposes; file fragments would return the filename, while other fragments would return some fixed identifier (much like type_name).

Reply:
That sounds good to me, but I think it should be called `ToString()`